packages/animations/browser/src/render/animation_renderer.ts

```diff
@@ -0,0 +1,134 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {AnimationTriggerMetadata} from '@angular/animations';
+import type {NgZone, Renderer2, RendererFactory2, RendererType2} from '@angular/core';
+
+import {AnimationEngine} from './animation_engine_next';
+import {AnimationRenderer, BaseAnimationRenderer} from './renderer';
+
+// Define a recursive type to allow for nested arrays of `AnimationTriggerMetadata`. Note that an
+// interface declaration is used as TypeScript prior to 3.7 does not support recursive type
+// references, see https://github.com/microsoft/TypeScript/pull/33050 for details.
+type NestedAnimationTriggerMetadata = AnimationTriggerMetadata|RecursiveAnimationTriggerMetadata;
+interface RecursiveAnimationTriggerMetadata extends Array<NestedAnimationTriggerMetadata> {}
+
+export class AnimationRendererFactory implements RendererFactory2 {
+  private _currentId: number = 0;
+  private _microtaskId: number = 1;
+  private _animationCallbacksBuffer: [(e: any) => any, any][] = [];
+  private _rendererCache = new Map<Renderer2, BaseAnimationRenderer>();
+  private _cdRecurDepth = 0;
+
+  constructor(
+      private delegate: RendererFactory2, private engine: AnimationEngine, private _zone: NgZone) {
+    engine.onRemovalComplete = (element: any, delegate: Renderer2) => {
+      // Note: if a component element has a leave animation, and a host leave animation,
+      // the view engine will call `removeChild` for the parent
+      // component renderer as well as for the child component renderer.
+      // Therefore, we need to check if we already removed the element.
+      const parentNode = delegate?.parentNode(element);
+      if (parentNode) {
+        delegate.removeChild(parentNode, element);
+      }
+    };
+  }
+
+  createRenderer(hostElement: any, type: RendererType2): Renderer2 {
+    const EMPTY_NAMESPACE_ID = '';
+
+    // cache the delegates to find out which cached delegate can
+    // be used by which cached renderer
+    const delegate = this.delegate.createRenderer(hostElement, type);
+    if (!hostElement || !type?.data?.['animation']) {
+      let renderer: BaseAnimationRenderer|undefined = this._rendererCache.get(delegate);
+      if (!renderer) {
+        // Ensure that the renderer is removed from the cache on destroy
+        // since it may contain references to detached DOM nodes.
+        const onRendererDestroy = () => this._rendererCache.delete(delegate);
+        renderer =
+            new BaseAnimationRenderer(EMPTY_NAMESPACE_ID, delegate, this.engine, onRendererDestroy);
+        // only cache this result when the base renderer is used
+        this._rendererCache.set(delegate, renderer);
+      }
```

Comment:
nit: to improve minification (property names are not minified), we can reference `this._rendererCache` once and use a const after that.
```suggestion
      const cache = this._rendererCache;
      let renderer: BaseAnimationRenderer|undefined = cache.get(delegate);
      if (!renderer) {
        // Ensure that the renderer is removed from the cache on destroy
        // since it may contain references to detached DOM nodes.
        const onRendererDestroy = () => cache.delete(delegate);
        renderer =
            new BaseAnimationRenderer(EMPTY_NAMESPACE_ID, delegate, this.engine, onRendererDestroy);
        // only cache this result when the base renderer is used
        cache.set(delegate, renderer);
      }
```

Reply:
Interesting knowledge, TIL 👍